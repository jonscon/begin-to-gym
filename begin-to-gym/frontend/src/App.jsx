import { useState, useEffect } from 'react'
import { BrowserRouter } from "react-router-dom";
import useLocalStorage from "./hooks/useLocalStorage";

import NavBar from "./routes-nav/NavBar.jsx";
import ReactRoutes from "./routes-nav/Routes.jsx";
import LoadingSpinner from "./general/LoadingSpinner";

import UserContext from './context/UserContext.jsx';
import { decodeToken } from "react-jwt";
import GymApi from "./api";
import "./App.css";

// Key name for localStorage - used for "remember me" when logging back in
export const TOKEN_STORAGE_ID = "gym-token";

/** Begin to Gym Application.
 * 
 * - currentUser: user object from API. Used to tell if someone
 *   is logged in. This is ppassed around through context
 *   throughout app.
 * 
 * - token: for logged in users, this is their authentication JWT.
 *   Required to be set for most API ccalls. This is initially read 
 *   from localStorage and synced to there via useLocalStorage hook.
 */

function App() {
  const [infoLoaded, setInfoLoaded] = useState(false);
  const [currentUser, setCurrentUser] = useState(null);
  const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);

  useEffect(() => {
    async function getCurrentUser() {
      if (token) {
        try {
          let { username } = decodeToken(token);
          // set token on Api calls so it can be used to call API.
          GymApi.token = token;
          let currentUser = await GymApi.getCurrentUser(username);
          setCurrentUser(currentUser);
        } catch (err) {
          console.error("App loadUserInfo: problem loading", err);
          setCurrentUser(null);
        }
      }
      setInfoLoaded(true);
    }
    setInfoLoaded(false);
    getCurrentUser();
  }, [token]);

  /** Handles signup. */
  async function signup(data) {
    try {
      let token = await GymApi.signup(data);
      setToken(token);
      return { success: true };
    } catch (err) {
      console.error("signup failed", err);
      return { success: false, err };
    }
  }

  /** Handles login. */
  async function login(data) {
    try {
      let token = await GymApi.login(data);
      setToken(token);
      return { success: true };
    } catch (err) {
      console.error("login failed", err);
      return { success: false, err };
    }
  }

  /** Handles logout. */
  function logout() {
    setCurrentUser(null);
    setToken(null);
  }

  if (!infoLoaded) return <LoadingSpinner />

  return (
      <BrowserRouter>
        <UserContext.Provider
            value={{ currentUser, setCurrentUser }}>
          <div className="App">
            <NavBar logout={logout} />
            <ReactRoutes login={login} signup={signup} />
          </div>
        </UserContext.Provider>
      </BrowserRouter>
  );
}

export default App;
