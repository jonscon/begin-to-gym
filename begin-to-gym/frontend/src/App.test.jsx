import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import App from "./App.jsx";

describe("test App.jsx", function () {
    it("renders without crashing", function () {
        render(<App />);
    })

    it("matchces snapshot", function () {
        const {asFragment} = render(<App />);
        expect(asFragment()).toMatchSnapshot();
    })
})
