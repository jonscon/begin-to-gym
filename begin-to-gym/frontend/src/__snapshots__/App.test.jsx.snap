// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test App.jsx > matchces snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="Navigation navbar navbar-expand-md navbar"
    >
      <div
        class="container-fluid"
      >
        <a
          aria-current="page"
          class="navbar-brand active"
          href="/"
        >
          Begin to Gym
        </a>
        <ul
          class="navbar-nav ml-auto nav"
        >
          <li
            class="nav-item mr-4 nav-item"
          >
            <a
              class="nav-link"
              href="/login"
            >
              Login
            </a>
          </li>
          <li
            class="nav-item mr-4 nav-item"
          >
            <a
              class="nav-link"
              href="/signup"
            >
              Sign Up
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div>
      <div
        class="Home"
      >
        <div
          class="container text-center"
        >
          <header>
            <h1>
              Begin to Gym
            </h1>
            <p
              class="lead"
            >
              A solution for your fitness goals!
            </p>
          </header>
          <span>
            <div
              class="Home-mission"
            >
              <h3>
                Our Mission
              </h3>
              <p>
                Life can be hectic, which can make working out seem daunting in your already-busy schedule. At Begin to Gym, our mission is simple: to help you simplify your fitness life. We'll simplify your goals by curating programs that fit your needs so that you don't have to!
              </p>
            </div>
            <p
              class="lead"
            >
              Get Started Below!
            </p>
            <a
              class="btn btn-primary font-weight-bold me-3"
              href="/login"
            >
              Log In
            </a>
            <a
              class="btn btn-primary font-weight-bold ms-3"
              href="/signup"
            >
              Sign Up
            </a>
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
